deadlines/src/main/java/com/palantir/deadlines/Deadlines.java

```diff
@@ -155,44 +177,70 @@ public static <T> void encodeToRequest(
      * {@link #getRemainingDeadline()}} from threads participating in the current trace. The deadline value
      * is read from a {@link DeadlinesHttpHeaders#EXPECT_WITHIN} header on the request object.
      *
+     * Enforcement of the deadline is enabled in onw of two ways:
+     *   - Callers can set the "internalEnforced" parameter on this method to true.
+     *   - If an "Expect-Within-Enforced" header is present in the request with a value of "true",
+     *     it will enable enforcement.
+     * If either of those conditions are met, the deadline state stored in the TraceLocal will additionally
+     * set an enforcement flag. This will indicate to future calls to {@link #encodeToRequest} that it should check
+     * for and enforce expiration. It also causes those calls to further propagate an "Expect-Within-Enforced" header,
+     * so that downstream receivers will also be required to enforce the deadline expiration.
+     *
      * This function has side-effects on the internal deadline state stored in a TraceLocal; the state is
      * set (or overwritten) based on the value of the deadline parsed from request headers.
      *
      * @param internalDeadline if present, represents an alternative deadline that should be used if it is
      * lower than the one parsed from a request header
      * @param request the request object to read the deadline value from
      * @param adapter a {@link RequestDecodingAdapter} that handles reading the header value from the request object
+     * @param internalEnforced whether this deadline should be enforced at this hop (and all downstream receivers)
      */
     public static <T> void parseFromRequest(
-            Optional<Duration> internalDeadline, T request, RequestDecodingAdapter<? super T> adapter) {
+            Optional<Duration> internalDeadline,
+            T request,
+            RequestDecodingAdapter<? super T> adapter,
+            boolean internalEnforced) {
```

Comment:
after chatting offline, i think we're gravitating towards making this a tri-state flag instead, where the states are:
- `enforce` == deadline expirations will be enforced by this hop, and the enforcement flag will be propagated
- `enforce_if_requested` == deadline expirations will be enforced by this hop, and the enforcement flag will be propagated, only if we see an `Expect-Within-Enforced: true` header; but otherwise will be ignored
- `ignore` == deadline expirations will be ignored by this hop, even if we see an `Expect-Within-Enforced: true` header, and the flag will _not_ be propagated (essentially terminates enforcement for this and all future hops)